internal/RF_Base.ecl

```diff
@@ -21,36 +22,134 @@ TreeNodeDat := Types.TreeNodeDat;
+  *    seems to provide strong benefits without significant down-side.
+  * This base module provides the common methods for building and using the forest model,
+  * while the derived modules (e.g. ClassificationForest and RegressionForest) overlay
+  * virtual attributes that perform work specific to the type of forest.
+  * The basic steps in building the tree are:
+  * - Organize the input data:
+  *   - Identify the work-items and number of training records and features
+  *   - Ensure that the record ids are sequential
+  *   - Create a map of each of the feature numbers for each work-item
+  * - Generate random bootstrap samples for each tree
+  * - Create numTree tree root nodes
+  * - Replicate the training data bootstraps to each tree root
+  * - Build the trees, one layer at a time by:
+  *   - Randomly choose the featuresPerNode features to evaluate for a given level
+  *   - Evaluate all possible splits on those features to find the feature and split-point
+  *     that results in the most pure pair of sub-groups.  The criteria for evaluating
+  *     splits is different for the different forest types.
+  *   - Choose the best split, and create a split-node (branch) based on that split.
+  *   - Create a left and right child node at the next level under the split-node.
+  *   - Reallocate the data points from the node to the left or right child nodes.
+  *   - If any group is pure (i.e. same dependent value for all members) or we've reached
+  *     maxDepth, replace all the datapoints in that node with a single leaf node.
+  * - At the end of the process, all the datapoints will have been absorbed by leaf nodes,
+  *   and the resulting trees will contain only split-nodes and leaf-nodes.
+  * - Note that a single common data structure (see LT_Types.TreeNodeData) is used to
+  *   represent:
+  *   - Data-points that need to be processed for a node (only during tree growing)
+  *   - A split-node
+  *   - A leaf-node
+  * - The resulting tree-nodes representing the forest are converted to a Forest Model
+  *   for later use in prediction or analysis.
   */
 EXPORT RF_Base(DATASET(GenField) X_in=DATASET([], GenField),
               DATASET(GenField) Y_In=DATASET([], GenField),
               UNSIGNED numTrees=100,
               UNSIGNED featuresPerNodeIn=0,
               UNSIGNED maxDepth=255) := MODULE
+  // Resequence ids to go from 1-numRecords.
+  SHARED resequenceByWi(DATASET(GenField) dat) := FUNCTION
+    // Dat should be evenly distributed at this point
+    xGen := RECORD(GenField)
+      t_RecordId newId := 0;
+    END;
+    xDat := PROJECT(dat, TRANSFORM(xGen, SELF := LEFT));
+    xGen setNewIds(xGen l, xGen r) := TRANSFORM
+      newId := IF(l.wi != r.wi, 1, IF(r.id = l.id, l.newId, l.newId + 1));
+      SELF.newId := newId;
+      SELF := r;
+    END;
+    xDatS := SORT(xDat, wi, id);
+    xDat2 := ITERATE(xDatS, setNewIds(LEFT, RIGHT));
```

Comment:
Comment only.  This looks like a case where PROCESS(...) would have been a better choice.  You are adding a field (newId) for the iteration and then removing it.  In this case the efficiency is not impacted because the field being manipulated is only a full word.  In more interesting cases though, the field could be a dataset used to implement a stack or queue.